File: scm-java/src/main/scheme/primitives/PrimitiveDirectoryEntries.java
package scheme.primitives;

import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;

import scheme.*;

public class PrimitiveDirectoryEntries extends Primitive {
    @Override
    public String name() {
        return "directory-entries";
    }

    @Override
    public String info() {
        return "Syntax: (directory-entries dirname)\n" +
               "Library: (scm fs)\n" +
               "Description: Returns a list of (name . type) pairs for the entries in dirname, where name is the entry name (not a full path) and type is one of the symbols file, directory, or symlink. Symlinks are reported as symlink regardless of what they point to (they are not followed).\n" +
               "Example:\n" +
               "  (directory-entries \"/tmp\") => ((\"a.txt\" . file) (\"sub\" . directory) (\"link\" . symlink) ...)";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);
        var dir = new String(Value.asString(arguments[0]));
        Object result = Value.NIL;
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(LongPath.of(dir))) {
            // Collect first so we can preserve directory order in the list.
            java.util.ArrayList<Path> entries = new java.util.ArrayList<>();
            for (Path p : stream) entries.add(p);
            for (int i = entries.size() - 1; i >= 0; i--) {
                Path p = entries.get(i);
                String type;
                if (Files.isSymbolicLink(p)) type = "symlink";
                else if (Files.isDirectory(p)) type = "directory";
                else type = "file";
                String entryName = p.getFileName().toString();
                result = new Pair(new Pair(entryName.toCharArray(), Value.intern(type)), result);
            }
            return result;
        } catch (Exception e) {
            return result;
        }
    }
}

File: scm-csharp/scheme/primitives/PrimitiveDirectoryEntries.cs
using System;
using System.IO;

namespace scheme;

public class PrimitiveDirectoryEntries : Primitive
{
    public override string Name()
    {
        return "directory-entries";
    }

    public override string Info()
    {
        return
            "Syntax: (directory-entries dirname)\n" +
            "Library: (scm fs)\n" +
            "Description: Returns a list of (name . type) pairs for the entries in dirname, where name is the entry name (not a full path) and type is one of the symbols file, directory, or symlink. Symlinks are reported as symlink regardless of what they point to (they are not followed).\n" +
            "Example:\n" +
            "  (directory-entries \"/tmp\") => ((\"a.txt\" . file) (\"sub\" . directory) (\"link\" . symlink) ...)";
    }

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        var dir = LongPath.Wlp(new String(Value.AsString(arguments[0])));
        var di = new DirectoryInfo(dir);
        var entries = di.GetFileSystemInfos();
        object result = Value.NIL;
        for (int i = entries.Length - 1; i >= 0; i--)
        {
            var e = entries[i];
            string type;
            if (e.LinkTarget != null) type = "symlink";
            else if ((e.Attributes & FileAttributes.Directory) != 0) type = "directory";
            else type = "file";
            result = new Pair(new Pair(e.Name.ToCharArray(), Value.Intern(type)), result);
        }
        return result;
    }
}
